comms/dht/src/store_forward/database/mod.rs

```diff
@@ -259,13 +263,30 @@ impl StoreAndForwardDatabase {
         since: NaiveDateTime,
     ) -> Result<usize, StorageError>
     {
+        let storage_capacity = self.storage_capacity;
         self.connection
             .with_connection_async(move |conn| {
-                diesel::delete(stored_messages::table)
+                let mut num_removed = diesel::delete(stored_messages::table)
                     .filter(stored_messages::stored_at.lt(since))
                     .filter(stored_messages::priority.eq(priority as i32))
-                    .execute(conn)
-                    .map_err(Into::into)
+                    .execute(conn)?;
+
+                let msg_count = stored_messages::table
+                    .select(dsl::count(stored_messages::id))
+                    .first::<i64>(conn)? as usize;
+                if msg_count > storage_capacity {
+                    let remove_count = msg_count - storage_capacity;
+                    let message_ids: Vec<i32> = stored_messages::table
+                        .select(stored_messages::id)
+                        .order_by(stored_messages::stored_at.asc())
+                        .limit(remove_count as i64)
+                        .get_results(conn)?;
+                    num_removed += diesel::delete(stored_messages::table)
+                        .filter(stored_messages::id.eq_any(message_ids))
+                        .execute(conn)?;
+                }
```

Comment:
I would do this in a new function `truncate_messages(max_size: usize)` and call it from the saf service - there's no extra query cost and it's better for a function to do one thing

Reply:
Also can you not do ordering and limit with a delete?

`DELETE FROM stored_messages ORDER BY stored_at ASC LIMIT {remove_count}`

Reply:
I changed it to make use of a `truncate_messages` function. I also attempted to perform the ordering and limit delete in one query, but was unable to get it to work.